File: 백준/Gold/16234. 인구 이동/인구 이동.py
from collections import deque

def bfs(a,b):
    quene=deque()
    stack=[]
    stack.append((a,b))
    quene.append((a,b))
    visited[a][b]=1
    total=0
    total+=mapping[a][b]
    while quene:
        i,j=quene.popleft()
        for k in range(4):
            ni,nj = i+di[k],j+dj[k]
            if 0<=ni<N and 0<=nj<N and L<=abs(mapping[i][j]-mapping[ni][nj])<=R and visited[ni][nj]==0:
                visited[ni][nj]=1
                total+=mapping[ni][nj]
                stack.append((ni,nj))
                quene.append((ni,nj))

    if len(stack)!=1:
        total=total//len(stack)
        for a,b in stack:
            mapping[a][b]=total



di=[0,0,1,-1]
dj=[1,-1,0,0]

N,L,R = map(int,input().split())
mapping=[list(map(int,input().split())) for _ in range(N)]

result=0
is_true = True
while True:
    count=0
    visited=[[0]*N for _ in range(N)]
    for i in range(N):
        for j in range(N):
            if visited[i][j]==0:
                bfs(i, j)
                count+=1

    if count==N*N:
        break
    result+=1

print(result)

File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static int N;
    static int L;
    static int R;
    static int[][] mapping;
    static boolean[][] visited;
    static int[][] dir = {{0,1},{0,-1},{1,0},{-1,0}};

    static void bfs(int a, int b) {
        Queue<int[]> quene = new LinkedList<>();
        Stack<int[]> stack = new Stack<>();
        quene.add(new int[]{a,b});
        stack.add(new int[]{a,b});
        int total=0;
        visited[a][b]=true;
        total += mapping[a][b];
        while (!quene.isEmpty()) {
            int[] now = quene.poll();
            int nowA = now[0];
            int nowB = now[1];
            for (int k=0;k< dir.length;k++) {
                int ni = nowA + dir[k][0];
                int nj = nowB + dir[k][1];
                if (0<=ni && ni<N && 0<=nj && nj<N && L<=Math.abs(mapping[nowA][nowB]-mapping[ni][nj]) && Math.abs(mapping[nowA][nowB]-mapping[ni][nj])<=R && !visited[ni][nj]) {
                    visited[ni][nj]=true;
                    total += mapping[ni][nj];
                    stack.add(new int[]{ni,nj});
                    quene.add(new int[]{ni,nj});
                }
            }
        }

        if (stack.size()!=1) {
            total=total/stack.size();
            for (int l=0;l<stack.size();l++) {
                int x = stack.get(l)[0];
                int y = stack.get(l)[1];
                mapping[x][y]=total;
            }
        }

    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N=Integer.parseInt(st.nextToken());
        L=Integer.parseInt(st.nextToken());
        R=Integer.parseInt(st.nextToken());
        mapping = new int[N][N];
        for (int i=0;i<N;i++) {
            st = new StringTokenizer(br.readLine());
            for (int j=0;j<N;j++) {
                mapping[i][j]=Integer.parseInt(st.nextToken());
            }
        }
        int result =0;
        while (true) {
            int count=0;
            visited = new boolean[N][N];
            for (int i=0;i<N;i++) {
                for (int j=0;j<N;j++) {
                    if (!visited[i][j]) {
                        bfs(i,j);
                        count++;

                    }
                }
            }
            if (count==N*N) break;

            result++;
        }
        System.out.println(result);
    }
}

